libwasmvm/src/api.rs

```diff
@@ -254,8 +261,28 @@ impl BackendApi for GoApi {
             Err(e) => return (Err(BackendError::unknown(e.to_string())), gas_info),
         };
         callee_instance.env.set_serialized_env(&contract_env);
-        callee_instance.set_storage_readonly(caller_env.is_storage_readonly());
         gas_info.cost += instantiate_cost;
+        // set read-write permission to callee instance
+        let is_read_write_permission =
+            match get_read_write_permission(&mut callee_instance, func_info) {
+                Ok(permission) => permission,
+                Err(e) => return (Err(e), gas_info),
+            };
+        if caller_env.is_storage_readonly() {
```

Comment:
I think it would be better to create a function that handles permissions with `caller_env` as an argument. Then this change to `fn contract_call` will be a single line.